In [152]:
#Load required packages
import pandas as pd
import numpy as np 
import researchpy as rp
import scipy.stats as stats

### AFL

Games are from the 2009 season til the 2020 season and the data is sourced from http://www.aussportsbetting.com/data/historical-afl-results-and-odds-data/

In [168]:
#Read in data and neccessary columns
data_afl = pd.read_excel("afl.xlsx")
data_afl.columns = data_afl.iloc[0]
data_afl = data_afl.drop(0)
data_afl = data_afl[['Home Team','Away Team','Home Score','Away Score','Home Odds','Away Odds']]

In [224]:
#Create new variables for our analysis
data_afl = data_afl[(data_afl['Home Score'] - data_afl['Away Score']) != 0] #filter out draws/no results
data_afl['Winner'] = np.where(data_afl['Home Score'] > data_afl['Away Score'],data_afl['Home Team'],data_afl['Away Team'])
data_afl['Underdog'] = np.where(data_afl['Home Odds'] < data_afl['Away Odds'],data_afl['Away Team'],data_afl['Home Team'])
data_afl['Bookmakers Margin'] = (1/data_afl['Home Odds'])+(1/data_afl['Away Odds']) - 1
data_afl['Home Implied Odds'] = 1/data_afl['Home Odds']
data_afl['Away Implied Odds'] = 1/data_afl['Away Odds']
data_afl['Home Fair Odds'] = data_afl['Home Implied Odds']/(data_afl['Bookmakers Margin']+1)
data_afl['Away Fair Odds'] = data_afl['Away Implied Odds']/(data_afl['Bookmakers Margin']+1)
data_afl['Home Expected Return'] = data_afl['Home Odds']*data_afl['Home Fair Odds'] - 1
data_afl['Away Expected Return'] = data_afl['Away Odds']*data_afl['Away Fair Odds'] - 1
data_afl['Home Actual Return'] = np.where(data_afl['Winner'] == data_afl['Home Team'], data_afl['Home Odds'] - 1, -1)
data_afl['Away Actual Return'] = np.where(data_afl['Winner'] == data_afl['Away Team'], data_afl['Away Odds'] - 1, -1)
data_afl['Underdog Actual Return'] = np.where(data_afl['Underdog'] == data_afl['Away Team'], data_afl['Away Actual Return'], data_afl['Home Actual Return'])
data_afl['Favourite Actual Return'] = np.where(data_afl['Underdog'] == data_afl['Home Team'], data_afl['Away Actual Return'], data_afl['Home Actual Return'])

In [190]:
#inspect dataframe
data_afl

0            Home Team         Away Team Home Score Away Score Home Odds  \
1             Richmond           Geelong         81         50      1.83   
2             Brisbane           Geelong         42         82      1.79   
3        Port Adelaide          Richmond         40         46      1.73   
4              Geelong       Collingwood        100         32       1.6   
5             Richmond          St Kilda         80         49      1.26   
6           West Coast       Collingwood         75         76      1.41   
7             St Kilda  Western Bulldogs         67         64      2.17   
8             Brisbane          Richmond         69         54      2.06   
9        Port Adelaide           Geelong         58         42      1.72   
10         Collingwood     Port Adelaide         45         61      2.29   
11           Fremantle  Western Bulldogs         44         74      3.25   
12              Sydney           Geelong         63         69      5.85   
13            Hawthorn        Gold Coast        108         57      2.05   
14            Brisbane           Carlton         78         61      1.18   
15            Adelaide          Richmond         33         77      4.31   
16            Essendon         Melbourne         49         68      2.56   
17            St Kilda        GWS Giants         82         30      1.78   
18     North Melbourne        West Coast         34         49      7.62   
19         Collingwood        Gold Coast         68         46      1.23   
20              Sydney          Brisbane         41         73      3.34   
21            Hawthorn  Western Bulldogs         40         76      3.84   
22             Carlton          Adelaide         56         72      1.32   
23          GWS Giants         Melbourne         74         79      1.68   
24       Port Adelaide          Essendon         79         29      1.13   
25     North Melbourne         Fremantle         35         99      2.68   
26             Geelong          Richmond         31         57      1.72   
27            St Kilda        West Coast         50         65      1.69   
28            Brisbane        Gold Coast         88         43      1.29   
29             Carlton            Sydney         57         52      1.54   
30            Adelaide        GWS Giants         59         47      3.63   
...                ...               ...        ...        ...       ...   
2262           Geelong         Melbourne        117         71      1.04   
2263           Carlton            Sydney        124         63      1.23   
2264          Essendon  Western Bulldogs         70        103       3.1   
2265         Melbourne     Port Adelaide        101         90      2.89   
2266          Hawthorn   North Melbourne         73         64       1.5   
2267        West Coast          St Kilda         70         90       4.7   
2268          Adelaide         Fremantle        130         13      1.14   
2269          Brisbane           Geelong        108         65      2.28   
2270           Carlton          Richmond        109         89      1.25   
2271            Sydney          Essendon         72        107      1.66   
2272  Western Bulldogs       Collingwood        110        111      1.56   
2273          St Kilda           Geelong         91         85       2.3   
2274         Fremantle           Carlton        100        115      2.28   
2275            Sydney   North Melbourne        100         85      1.31   
2276  Western Bulldogs          Hawthorn        133         45       1.4   
2277          Richmond          Adelaide         85        102      3.78   
2278     Port Adelaide          Brisbane        128         80       2.3   
2279         Melbourne        West Coast        112         92      1.96   
2280       Collingwood          Essendon        102         67       1.5   
2281          St Kilda          Richmond         92         36      1.07   
2282   North Melbourne  Western Bulldogs        

### NRL

Games are from the 2009 season til the 2020 season and the data is sourced from http://www.aussportsbetting.com/data/historical-nrl-results-and-odds-data/

In [225]:
#Read in data and neccessary columns
data_nrl = pd.read_excel("nrl.xlsx")
data_nrl.columns = data_nrl.iloc[0]
data_nrl = data_nrl.drop(0)
print(data_nrl.columns)
data_nrl = data_nrl[['Home Team','Away Team','Home Score','Away Score','Home Odds','Away Odds']]
data_nrl = data_nrl.dropna()

Index(['Date', 'Kick-off (local)', 'Home Team', 'Away Team', 'Home Score',
       'Away Score', 'Play Off Game?', 'Over Time?', 'Home Odds', 'Draw Odds',
       'Away Odds', 'Bookmakers Surveyed', 'Home Odds Open', 'Home Odds Min',
       'Home Odds Max', 'Home Odds Close', 'Away Odds Open', 'Away Odds Min',
       'Away Odds Max', 'Away Odds Close', 'Home Line Open', 'Home Line Min',
       'Home Line Max', 'Home Line Close', 'Away Line Open', 'Away Line Min',
       'Away Line Max', 'Away Line Close', 'Home Line Odds Open',
       'Home Line Odds Min', 'Home Line Odds Max', 'Home Line Odds Close',
       'Away Line Odds Open', 'Away Line Odds Min', 'Away Line Odds Max',
       'Away Line Odds Close', 'Total Score Open', 'Total Score Min',
       'Total Score Max', 'Total Score Close', 'Total Score Over Open',
       'Total Score Over Min', 'Total Score Over Max',
       'Total Score Over Close', 'Total Score Under Open',
       'Total Score Under Min', 'Total Score Under Max',
      

In [226]:
#Create new variables for our analysis
data_nrl = data_nrl[(data_nrl['Home Score'] - data_nrl['Away Score']) != 0] #filter out draws/no results
data_nrl['Winner'] = np.where(data_nrl['Home Score'] > data_nrl['Away Score'],data_nrl['Home Team'],data_nrl['Away Team'])
data_nrl['Underdog'] = np.where(data_nrl['Home Odds'] < data_nrl['Away Odds'],data_nrl['Away Team'],data_nrl['Home Team'])
data_nrl['Bookmakers Margin'] = (1/data_nrl['Home Odds'])+(1/data_nrl['Away Odds']) - 1
data_nrl['Home Implied Odds'] = 1/data_nrl['Home Odds']
data_nrl['Away Implied Odds'] = 1/data_nrl['Away Odds']
data_nrl['Home Fair Odds'] = data_nrl['Home Implied Odds']/(data_nrl['Bookmakers Margin']+1)
data_nrl['Away Fair Odds'] = data_nrl['Away Implied Odds']/(data_nrl['Bookmakers Margin']+1)
data_nrl['Home Expected Return'] = data_nrl['Home Odds']*data_nrl['Home Fair Odds'] - 1
data_nrl['Away Expected Return'] = data_nrl['Away Odds']*data_nrl['Away Fair Odds'] - 1
data_nrl['Home Actual Return'] = np.where(data_nrl['Winner'] == data_nrl['Home Team'], data_nrl['Home Odds'] - 1, -1)
data_nrl['Away Actual Return'] = np.where(data_nrl['Winner'] == data_nrl['Away Team'], data_nrl['Away Odds'] - 1, -1)
data_nrl['Underdog Actual Return'] = np.where(data_nrl['Underdog'] == data_nrl['Away Team'], data_nrl['Away Actual Return'], data_nrl['Home Actual Return'])
data_nrl['Favourite Actual Return'] = np.where(data_nrl['Underdog'] == data_nrl['Home Team'], data_nrl['Away Actual Return'], data_nrl['Home Actual Return'])

In [227]:
#inspect dataframe
data_nrl

0                         Home Team                      Away Team Home Score  \
1                  Penrith Panthers                Melbourne Storm         20   
2                  Penrith Panthers         South Sydney Rabbitohs         20   
3                   Melbourne Storm               Canberra Raiders         30   
4                   Parramatta Eels         South Sydney Rabbitohs         24   
5                   Sydney Roosters               Canberra Raiders         18   
6            South Sydney Rabbitohs              Newcastle Knights         46   
7                   Melbourne Storm                Parramatta Eels         36   
8                  Canberra Raiders                Cronulla Sharks         32   
9                  Penrith Panthers                Sydney Roosters         29   
10                St George Dragons                Melbourne Storm         30   
11             New Zealand Warriors               Manly Sea Eagles         40   
12                     Wests Tigers                Parramatta Eels         24   
13                  Cronulla Sharks               Canberra Raiders         28   
14              Canterbury Bulldogs               Penrith Panthers          0   
15           South Sydney Rabbitohs                Sydney Roosters         60   
16                Gold Coast Titans              Newcastle Knights         36   
17                 Brisbane Broncos              North QLD Cowboys         16   
18                Newcastle Knights              St George Dragons         42   
19                 Canberra Raiders           New Zealand Warriors         26   
20                  Sydney Roosters                Cronulla Sharks         34   
21                  Melbourne Storm                   Wests Tigers         50   
22                 Manly Sea Eagles              Gold Coast Titans         24   
23                  Parramatta Eels               Brisbane Broncos         26   
24                North QLD Cowboys               Penrith Panthers         12   
25           South Sydney Rabbitohs            Canterbury Bulldogs         16   
26                  Cronulla Sharks           New Zealand Warriors         22   
27                  Melbourne Storm              North QLD Cowboys         36   
28                  Sydney Roosters              Newcastle Knights         42   
29                Gold Coast Titans               Brisbane Broncos         18   
30                St George Dragons               Canberra Raiders          8   
...                             ...                            ...        ...   
2351  Canterbury-Bankstown Bulldogs     Cronulla-Sutherland Sharks         24   
2352           New Zealand Warriors         South Sydney Rabbitohs         16   
2353                Melbourne Storm              Gold Coast Titans          6   
2354               Penrith Panthers                   Wests Tigers         42   
2355                Sydney Roosters                Parramatta Eels         24   
2356               Brisbane Broncos   St. George Illawarra Dragons         12   
2357     Manly-Warringah Sea Eagles               Penrith Panthers         10   
2358   St. George Illawarra Dragons     Cronulla-Sutherland Sharks         10   
2359         South Sydney Rabbitohs              Newcastle Knights         22   
2360       North Queensland Cowboys                Melbourne Storm         12   
2361                Parramatta Eels               Canberra Raiders         18   
2362           New Zealand Warriors               Brisbane Broncos         10   
2363              Gold Coast Titans  Canterbury-Bankstown Bulldogs         20   
2364                   Wests Tigers                Sydney Roosters         40   
2365     Cronulla-Sutherland Sharks              Newcastle Knights         12   
2366     Manly-Warringah Sea Eagles           New Zealand Warriors         24   
2367               Canberra Raiders                Sydney Roosters          4   
2368       North Queensland Cow

### NFL

Games are from the 2006 season til the 2020 season and the data is sourced from http://www.aussportsbetting.com/data/historical-nfl-results-and-odds-data/

In [228]:
#Read in data and neccessary columns
data_nfl = pd.read_excel("nfl.xlsx")
print(data_nfl.columns)
data_nfl = data_nfl[['Home Team','Away Team','Home Score','Away Score','Home Odds Close','Away Odds Close']]
data_nfl = data_nfl.dropna()

Index(['Date', 'Home Team', 'Away Team', 'Home Score', 'Away Score',
       'Overtime?', 'Playoff Game?', 'Neutral Venue?', 'Home Odds Open',
       'Home Odds Min', 'Home Odds Max', 'Home Odds Close', 'Away Odds Open',
       'Away Odds Min', 'Away Odds Max', 'Away Odds Close', 'Home Line Open',
       'Home Line Min', 'Home Line Max', 'Home Line Close', 'Away Line Open',
       'Away Line Min', 'Away Line Max', 'Away Line Close',
       'Home Line Odds Open', 'Home Line Odds Min', 'Home Line Odds Max',
       'Home Line Odds Close', 'Away Line Odds Open', 'Away Line Odds Min',
       'Away Line Odds Max', 'Away Line Odds Close', 'Total Score Open',
       'Total Score Min', 'Total Score Max', 'Total Score Close',
       'Total Score Over Open', 'Total Score Over Min', 'Total Score Over Max',
       'Total Score Over Close', 'Total Score Under Open',
       'Total Score Under Min', 'Total Score Under Max',
       'Total Score Under Close', 'Notes'],
      dtype='object')


In [230]:
#Create new variables for our analysis
data_nfl = data_nfl[(data_nfl['Home Score'] - data_nfl['Away Score']) != 0]
data_nfl['Winner'] = np.where(data_nfl['Home Score'] > data_nfl['Away Score'],data_nfl['Home Team'],data_nfl['Away Team'])
data_nfl['Underdog'] = np.where(data_nfl['Home Odds Close'] < data_nfl['Away Odds Close'],data_nfl['Away Team'],data_nfl['Home Team'])
data_nfl['Bookmakers Margin'] = (1/data_nfl['Home Odds Close'])+(1/data_nfl['Away Odds Close']) - 1
data_nfl['Home Implied Odds'] = 1/data_nfl['Home Odds Close']
data_nfl['Away Implied Odds'] = 1/data_nfl['Away Odds Close']
data_nfl['Home Fair Odds'] = data_nfl['Home Implied Odds']/(data_nfl['Bookmakers Margin']+1)
data_nfl['Away Fair Odds'] = data_nfl['Away Implied Odds']/(data_nfl['Bookmakers Margin']+1)
data_nfl['Home Expected Return'] = data_nfl['Home Odds Close']*data_nfl['Home Fair Odds'] - 1
data_nfl['Away Expected Return'] = data_nfl['Away Odds Close']*data_nfl['Away Fair Odds'] - 1
data_nfl['Home Actual Return'] = np.where(data_nfl['Winner'] == data_nfl['Home Team'], data_nfl['Home Odds Close'] - 1, -1)
data_nfl['Away Actual Return'] = np.where(data_nfl['Winner'] == data_nfl['Away Team'], data_nfl['Away Odds Close'] - 1, -1)
data_nfl['Underdog Actual Return'] = np.where(data_nfl['Underdog'] == data_nfl['Away Team'], data_nfl['Away Actual Return'], data_nfl['Home Actual Return'])
data_nfl['Favourite Actual Return'] = np.where(data_nfl['Underdog'] == data_nfl['Home Team'], data_nfl['Away Actual Return'], data_nfl['Home Actual Return'])

In [231]:
#inspect dataframe
data_nfl

Home Team                 Away Team  Home Score  \
0         Tampa Bay Buccaneers          Los Angeles Rams          24   
1            Las Vegas Raiders        Kansas City Chiefs          31   
2           Indianapolis Colts         Green Bay Packers          34   
3            Minnesota Vikings            Dallas Cowboys          28   
4               Denver Broncos            Miami Dolphins          20   
5         Los Angeles Chargers             New York Jets          34   
6             Baltimore Ravens          Tennessee Titans          24   
7            Carolina Panthers             Detroit Lions          20   
8             Cleveland Browns       Philadelphia Eagles          22   
9               Houston Texans      New England Patriots          27   
10        Jacksonville Jaguars       Pittsburgh Steelers           3   
11          New Orleans Saints           Atlanta Falcons          24   
12    Washington Football Team        Cincinnati Bengals          20   
13            Seattle Seahawks         Arizona Cardinals          28   
14               Chicago Bears         Minnesota Vikings          13   
15        New England Patriots          Baltimore Ravens          23   
16            Los Angeles Rams          Seattle Seahawks          23   
17          New Orleans Saints       San Francisco 49ers          27   
18           Arizona Cardinals             Buffalo Bills          32   
19           Las Vegas Raiders            Denver Broncos          37   
20              Miami Dolphins      Los Angeles Chargers          29   
21           Carolina Panthers      Tampa Bay Buccaneers          23   
22            Cleveland Browns            Houston Texans          10   
23               Detroit Lions  Washington Football Team          30   
24           Green Bay Packers      Jacksonville Jaguars          24   
25             New York Giants       Philadelphia Eagles          27   
26         Pittsburgh Steelers        Cincinnati Bengals          36   
27            Tennessee Titans        Indianapolis Colts          17   
28               New York Jets      New England Patriots          27   
29        Tampa Bay Buccaneers        New Orleans Saints           3   
...                        ...                       ...         ...   
1733            Denver Broncos        Kansas City Chiefs          24   
1734         Green Bay Packers             New York Jets          31   
1735           Oakland Raiders            Houston Texans          14   
1736        San Diego Chargers          Seattle Seahawks          30   
1737      Tampa Bay Buccaneers            St. Louis Rams          17   
1738             Buffalo Bills            Miami Dolphins          29   
1739         Carolina Panthers             Detroit Lions          24   
1740        Cincinnati Bengals           Atlanta Falcons          24   
1741          Cleveland Browns        New Orleans Saints          26   
1742         Minnesota Vikings      New England Patriots           7   
1743           New York Giants         Arizona Cardinals          14   
1744          Tennessee Titans            Dallas Cowboys          10   
1745       Washington Redskins      Jacksonville Jaguars          41   
1746          Baltimore Ravens       Pittsburgh Steelers          26   
1747         Arizona Cardinals        San Diego Chargers          18   
1748             Detroit Lions           New York Giants          35   
1749            Denver Broncos        Indianapolis Colts          31   
1750            Dallas Cowboys       San Francisco 49ers          17   
1751      Tampa Bay Buccaneers         Carolina Panthers          14   
1752           Atlanta Falcons        New Orleans Saints          37   
1753          Baltimore Ravens        Cincinnati Bengals          16   
1754             Chicago Bears             Buffalo Bills          20   
1755            Houston Texans       Washington Redskins          17   
1756        Kansas City Chiefs          Tennessee Titans     

### NBA

Includes all games from the 2008-09 season til the 2019-20 season and is sourced from https://www.sportsbookreviewsonline.com/scoresoddsarchives/nba/nbaoddsarchives.htm

In [232]:
#Read in data and neccessary columns
data_nba = pd.DataFrame()
nba = ["nba odds 2019-20.xlsx","nba odds 2018-19.xlsx","nba odds 2017-18.xlsx","nba odds 2016-17.xlsx",
       "nba odds 2015-16.xlsx","nba odds 2014-15.xlsx","nba odds 2013-14.xlsx","nba odds 2012-13.xlsx",
       "nba odds 2011-12.xlsx","nba odds 2010-11.xlsx","nba odds 2009-10.xlsx","nba odds 2008-09.xlsx"]

for season in nba:
    data = pd.read_excel(season)
    data = data[['Date','Team','ML','Final']]
    data['Odds'] = np.where(data['ML'] > 0, 1 + data['ML']/100, 1 - (100/data['ML']))
    data = data.drop(['Date','ML'], axis = 1)
    data_nba = pd.concat([data_nba,data])

In [233]:
#Reshape data to contain one game per row instead of just one team
home_team = data_nba.iloc[list(range(1,len(data_nba),2))]
away_team = data_nba.iloc[list(range(0,len(data_nba),2))]
home_team.columns = ['Home Team','Home Score','Home Odds']
away_team.columns = ['Away Team','Away Score','Away Odds']
home_team = home_team.reset_index(drop=True)
away_team = away_team.reset_index(drop=True)
data_nba = pd.concat([home_team,away_team], axis = 1)

In [234]:
#Create new variables for our analysis
data_nba['Winner'] = np.where(data_nba['Home Score'] > data_nba['Away Score'],data_nba['Home Team'],data_nba['Away Team'])
data_nba['Underdog'] = np.where(data_nba['Home Odds'] < data_nba['Away Odds'],data_nba['Away Team'],data_nba['Home Team'])
data_nba['Bookmakers Margin'] = (1/data_nba['Home Odds'])+(1/data_nba['Away Odds']) - 1
data_nba['Home Implied Odds'] = 1/data_nba['Home Odds']
data_nba['Away Implied Odds'] = 1/data_nba['Away Odds']
data_nba['Home Fair Odds'] = data_nba['Home Implied Odds']/(data_nba['Bookmakers Margin']+1)
data_nba['Away Fair Odds'] = data_nba['Away Implied Odds']/(data_nba['Bookmakers Margin']+1)
data_nba['Home Expected Return'] = data_nba['Home Odds']*data_nba['Home Fair Odds'] - 1
data_nba['Away Expected Return'] = data_nba['Away Odds']*data_nba['Away Fair Odds'] - 1
data_nba['Home Actual Return'] = np.where(data_nba['Winner'] == data_nba['Home Team'], data_nba['Home Odds'] - 1, -1)
data_nba['Away Actual Return'] = np.where(data_nba['Winner'] == data_nba['Away Team'], data_nba['Away Odds'] - 1, -1)
data_nba['Underdog Actual Return'] = np.where(data_nba['Underdog'] == data_nba['Away Team'], data_nba['Away Actual Return'], data_nba['Home Actual Return'])
data_nba['Favourite Actual Return'] = np.where(data_nba['Underdog'] == data_nba['Home Team'], data_nba['Away Actual Return'], data_nba['Home Actual Return'])

In [235]:
#inspect dataframe
data_nba

Home Team  Home Score  Home Odds     Away Team  Away Score  \
0           Toronto         130   1.357143    NewOrleans         122   
1        LAClippers         112   2.500000      LALakers         102   
2           Indiana         110   1.333333       Detroit         119   
3           Orlando          94   1.181818     Cleveland          85   
4         Charlotte         126   2.450000       Chicago         125   
5      Philadelphia         107   1.416667        Boston          93   
6             Miami         120   1.333333       Memphis         101   
7          Brooklyn         126   1.606061     Minnesota         127   
8        SanAntonio         120   1.133333       NewYork         111   
9            Dallas         108   1.222222    Washington         100   
10             Utah         100   1.200000  OklahomaCity          95   
11          Phoenix         124   1.952381    Sacramento          95   
12         Portland         100   2.000000        Denver         108   
13          Detroit         100   1.800000       Atlanta         117   
14          Houston         111   1.833333     Milwaukee         117   
15      GoldenState         122   1.909091    LAClippers         141   
16           Boston         112   1.714286       Toronto         106   
17        Charlotte          99   2.800000     Minnesota         121   
18         Brooklyn         113   1.277778       NewYork         109   
19     OklahomaCity          85   1.250000    Washington          97   
20       NewOrleans         116   1.769231        Dallas         123   
21          Memphis         102   2.050000       Chicago         110   
22           Denver         108   1.100000       Phoenix         107   
23       Sacramento         112   2.000000      Portland         122   
24         LALakers          95   1.625000          Utah          86   
25        Milwaukee         126   1.125000         Miami         131   
26          Detroit         111   2.600000  Philadelphia         117   
27          NewYork          95   2.900000        Boston         118   
28          Atlanta         103   2.150000       Orlando          99   
29        Cleveland         110   2.600000       Indiana          99   
...             ...         ...        ...           ...         ...   
15318        Dallas         105   1.588235        Denver         106   
15319       Atlanta          82   4.500000     Cleveland          97   
15320       Houston          99   4.250000      LALakers          87   
15321       Orlando          94   1.444444        Boston          95   
15322       Atlanta          74   5.500000     Cleveland          84   
15323        Dallas         119   1.714286        Denver         117   
15324        Boston          92   1.714286       Orlando          88   
15325      LALakers         118   1.071429       Houston          78   
15326        Denver         124   1.142857        Dallas         110   
15327       Orlando          83   1.312500        Boston          75   
15328       Houston          95   5.000000      LALakers          80   
15329        Boston          82   1.689655       Orlando         101   
15330      LALakers          89   1.080000       Houston          70   
15331      LALakers         105   1.357143        Denver         103   
15332     Cleveland         106   1.217391       Orlando         107   
15333      LALakers         103   1.416667        Denver         106   
15334     Cleveland          96   1.160000       Orlando          95   
15335        Denver          97   1.588235      LALakers         103   
15336       Orlando          99   1.833333     Cleveland          89   
15337        Denver         120   1.465116      LALakers         101   
15338       Orlando         116   2.000000     Cleveland         114   
15339      LALakers         103   1.377358        Denver          94   
15340     Cleveland         112   1.240964       Orlando         102   
15341        Denver          92   1.454545      LA

### Australia (AFL + NRL)

In [236]:
#Combine AFL and NRL games
australia = pd.concat([data_nrl, data_afl])

C:\Users\jamie\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


In [237]:
#inspect DataFrame
australia

Away Actual Return Away Difference Away Expected Return Away Fair Odds  \
1                   0.7             NaN           -0.0335025       0.568528   
2                    -1             NaN           -0.0376316       0.302632   
3                    -1             NaN            -0.037819       0.336427   
4                   0.3             NaN           -0.0470226        0.73306   
5                  1.91             NaN           -0.0411751       0.329493   
6                    -1             NaN           -0.0438351       0.270103   
7                    -1             NaN           -0.0460167       0.198333   
8                    -1             NaN           -0.0456726       0.191248   
9                    -1             NaN           -0.0429592       0.423469   
10                   -1             NaN           -0.0465796       0.467363   
11                   -1             NaN           -0.0404167       0.489583   
12                 0.41             NaN           -0.0428702       0.678815   
13                 1.55             NaN             -0.04375          0.375   
14                 0.16             NaN            -0.049595       0.819315   
15                   -1             NaN           -0.0418251       0.760456   
16                   -1             NaN           -0.0432099       0.617284   
17                 0.77             NaN           -0.0416321       0.541451   
18                   -1             NaN           -0.0475949       0.278481   
19                   -1             NaN           -0.0427876       0.152909   
20                   -1             NaN           -0.0466113       0.156293   
21                   -1             NaN           -0.0456396       0.155433   
22                 1.71             NaN           -0.0427685       0.353222   
23                   -1             NaN           -0.0511534       0.113228   
24                 0.15             NaN           -0.0419448       0.833091   
25                 5.72             NaN           -0.0547315       0.140665   
26                   -1             NaN            -0.041604       0.401003   
27                   -1             NaN           -0.0474256       0.109744   
28                   -1             NaN           -0.0438072       0.106362   
29                   -1             NaN           -0.0358897       0.408521   
30                 0.33             NaN           -0.0369917       0.724066   
...                 ...             ...                  ...            ...   
2262                 -1       -0.944198           -0.0558016      0.0921169   
2263                 -1       -0.935661           -0.0643385         0.2393   
2264               0.36        0.414709           -0.0547085       0.695067   
2265                 -1       -0.934005           -0.0659953       0.676815   
2266                 -1         -0.9375              -0.0625          0.375   
2267               0.17        0.233203           -0.0632027       0.800681   
2268                 -1       -0.942492           -0.0575076       0.173252   
2269                 -1       -0.929766           -0.0702338       0.592208   
2270                 -1       -0.938124           -0.0618762       0.249501   
2271               1.14         1.20516           -0.0651579       0.436842   
2272               1.34           1.404               -0.064            0.4   
2273                 -1       -0.929534           -0.0704663       0.595855   
2274               0.57        0.640234           -0.0702338       0.592208   
2275                 -1       -0.939352           -0.0606479       0.282937   
2276                 -1       -0.933333           -0.0666667       0.333333   
2277               0.25        0.310636           -0.0606362       0.751491   
2278                 -1       -0.933075           -0.0669251       0.594315   
2279                 -1       -0.924528           -0.0754717       0.528302   
2280                 -1         -0.9375              -0

In [188]:
#Total Underdog Actual Return
sum(australia['Underdog Actual Return'])

-456.8099999999998

In [238]:
#Total Favourite Actual Return
sum(australia['Favourite Actual Return'])

-161.9699999999998

In [147]:
#Expected Return - Note this is the same both underdog and favourite teams
sum(australia['Home Expected Return'])

-225.7188155658405

So here we are conducting a t-test to determine whether there is a difference in average return when betting on the favourite vs the underdog in Australia. Our hypotheseses are:

- Null Hypothesis: There is no difference in return between the Favourite teams and the Underdog teams

- Alternative Hypothesis: The Underdog teams produce less return than the Favourite teams

The assumptions are:

1) The data is continous - Which is True

2) Data follows the normal probability distribution - We can make this assumption has the sample size of n = 4637 is large enough such that the data follows the central limit theorem

3) The two samples are indepedendent - We have to be careful here as the underdog and favourite return are dependent on each other i.e. when the favourite wins the underdog loses and vice versa. To ensure we do not violate this assumption, we are going to compare the mean of the expected return and the mean of the underdog actual return 

4) The sample is random and observations are independent of each other - Which is true as the odds of each game are not dependent on the odds of another game

5) Variance of the two groups are equal - We cannot make this assumption as demonstrated by the Levene's test of equality of variance below 

In [240]:
#Determine whether we can assume equal variance
levene(australia['Underdog Actual Return'], australia['Home Expected Return'])

LeveneResult(statistic=1701.7961404106993, pvalue=0.0)

In [242]:
summary, results = rp.ttest(group1= australia['Underdog Actual Return'], group1_name= "Underdog",
         group2= australia['Home Expected Return'], group2_name= "Expected Return", equal_variances = False)

In [243]:
summary

Variable       N      Mean        SD        SE  95% Conf.  Interval
0         Underdog  4637.0 -0.098514  1.477552  0.021698  -0.141053 -0.055975
1  Expected Return  4637.0 -0.048678  0.008577  0.000126  -0.048925 -0.048431
2         combined  9274.0 -0.073596  1.045046  0.010852  -0.094868 -0.052324

In [244]:
results

Welch's t-test    results
0  Difference (Underdog - Expected Return) =     -0.0498
1                       Degrees of freedom =   4636.3125
2                                        t =     -2.2968
3                    Two side test p value =      0.0217
4                   Difference < 0 p value =      0.0108
5                   Difference > 0 p value =      0.9892
6                                Cohen's d =     -0.0477
7                                Hedge's g =     -0.0477
8                            Glass's delta =     -0.0337
9                                        r =      0.0337

With a p value of 0.0108 we reject the null hypothesis and accept the alternative hypothesis that in Australia, the Underdog teams produce less return than the Favourite teams 

### USA (NBA + NFL)

In [210]:
#combine nba and nfl games
usa = pd.concat([data_nba, data_nfl])

C:\Users\jamie\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  """Entry point for launching an IPython kernel.


In [211]:
#inspect dataframe
usa

Away Actual Return  Away Difference  Away Expected Return  \
0              -1.000000        -0.961656             -0.038344   
1              -1.000000        -0.958904             -0.041096   
2               2.400000         2.442254             -0.042254   
3              -1.000000        -0.955882             -0.044118   
4              -1.000000        -0.963583             -0.036417   
5              -1.000000        -0.951737             -0.048263   
6              -1.000000        -0.957746             -0.042254   
7               1.420000         1.454623             -0.034623   
8              -1.000000        -0.959368             -0.040632   
9              -1.000000        -0.961165             -0.038835   
10             -1.000000        -0.957983             -0.042017   
11             -1.000000        -0.955038             -0.044962   
12              0.833333         0.876812             -0.043478   
13              1.050000         1.091558             -0.041558   
14              1.000000         1.043478             -0.043478   
15              0.909091         0.954545             -0.045455   
16             -1.000000        -0.963504             -0.036496   
17              0.454545         0.497280             -0.042735   
18             -1.000000        -0.956236             -0.043764   
19              3.000000         3.047619             -0.047619   
20              1.100000         1.139761             -0.039761   
21              0.800000         0.841558             -0.041558   
22             -1.000000        -0.959302             -0.040698   
23              0.833333         0.876812             -0.043478   
24             -1.000000        -0.968944             -0.031056   
25              5.500000         5.540984             -0.040984   
26              0.526316         0.564581             -0.038265   
27              0.416667         0.464929             -0.048263   
28             -1.000000        -0.961923             -0.038077   
29             -1.000000        -0.961735             -0.038265   
...                  ...              ...                   ...   
1733           -1.000000        -0.976116             -0.023884   
1734           -1.000000        -0.976061             -0.023939   
1735            0.645000         0.668980             -0.023980   
1736           -1.000000        -0.975851             -0.024149   
1737            1.840000         1.864017             -0.024017   
1738           -1.000000        -0.974816             -0.025184   
1739           -1.000000        -0.974816             -0.025184   
1740           -1.000000        -0.976401             -0.023599   
1741           -1.000000        -0.976126             -0.023874   
1742            0.609000         0.633133             -0.024133   
1743            1.040000         1.064620             -0.024620   
1744            1.510000         1.533617             -0.023617   
1745           -1.000000        -0.976126             -0.023874   
1746           -1.000000        -0.976592             -0.023408   
1747           -1.000000        -0.975764             -0.024236   
1748           -1.000000        -0.975815             -0.024185   
1749           -1.000000        -0.976254             -0.023746   
1750            0.649000         0.673236             -0.024236   
1751            2.010000         2.034149             -0.024149   
1752           -1.000000        -0.975400             -0.024600   
1753            1.020000         1.045184             -0.025184   
1754            2.390000         2.414307             -0.024307   
1755           -1.000000        -0.975147             -0.024853   
1756            1.440000         1.464600             -0.024600   
1757           -1.000000        -0.975619             -0.024381   
1758           -1.000000        -0.975750             -0.024250   
1759           -1.000000        -0.976146             -0.023854   
1760           -1.000000        -0.976401            

In [245]:
#Total Underdog Actual Return
sum(usa['Underdog Actual Return'])

-657.0272390851745

In [247]:
#Total Favourite Actual Return
sum(usa['Favourite Actual Return'])

-686.877683331481

In [248]:
#Expected Return - Note this is the same both underdog and favourite teams
sum(usa['Home Expected Return'])

-616.7414972479099

Very similar to what we did with the Australia betting market, we are conducting a t-test to determine whether there is a difference in average return when betting on the favourite vs the underdog in the USA. Our hypotheseses are:

- Null Hypothesis: There is no difference in return between the Favourite teams and the Underdog teams

- Alternative Hypothesis: The Underdog teams produce different returns than the Favourite teams

The assumptions are:

1) The data is continous - Which is True

2) Data follows the normal probability distribution - We can make this assumption has the sample size of n = 17103 is large enough such that the data follows the central limit theorem

3) The two samples are indepedendent - We have to be careful here as the underdog and favourite return are dependent on each other i.e. when the favourite wins the underdog loses and vice versa. To ensure we do not violate this assumption, we are going to compare the mean of the expected return and the mean of the underdog actual return 

4) The sample is random and observations are independent of each other - Which is true as the odds of each game are not dependent on the odds of another game

5) Variance of the two groups are equal - We cannot make this assumption as demonstrated by the Levene's test of equality of variance below 

In [249]:
#Determine whether we can assume equal variance
levene(usa['Underdog Difference'], usa['Favourite Difference'])

LeveneResult(statistic=1030.0598728713067, pvalue=1.0684234288551446e-222)

In [250]:
summary, results = rp.ttest(group1= usa['Underdog Actual Return'], group1_name= "Underdog",
         group2= usa['Home Expected Return'], group2_name= "Expected Return", equal_variances = False)

In [251]:
summary

Variable        N      Mean        SD        SE  95% Conf.  Interval
0         Underdog  17103.0 -0.038416  1.644039  0.012571  -0.063057 -0.013775
1  Expected Return  17103.0 -0.036060  0.011625  0.000089  -0.036235 -0.035886
2         combined  34206.0 -0.037238  1.162524  0.006286  -0.049558 -0.024918

In [252]:
results

Welch's t-test     results
0  Difference (Underdog - Expected Return) =      -0.0024
1                       Degrees of freedom =   17103.7102
2                                        t =      -0.1874
3                    Two side test p value =       0.8514
4                   Difference < 0 p value =       0.4257
5                   Difference > 0 p value =       0.5743
6                                Cohen's d =      -0.0020
7                                Hedge's g =      -0.0020
8                            Glass's delta =      -0.0014
9                                        r =       0.0014

With a p value of 0.8514, we accept the null hypothesis that there is no difference in return when betting on the underdog as opposed to the favourite team in the USA.